db/snapshot/backup_task.cc

```diff
@@ -0,0 +1,129 @@
+    return tasks::is_internal::no;
+}
+
+static std::pair<sstring, table_id> extract_cf_name_and_uuid(const sstring& directory_name) {
+    // cf directory is of the form: 'cf_name-uuid'
+    // uuid is assumed to be exactly 32 hex characters wide.
+    constexpr size_t uuid_size = 32;
+    ssize_t pos = directory_name.size() - uuid_size - 1;
+    if (pos <= 0 || directory_name[pos] != '-') {
+        on_internal_error(snap_log, format("table directory entry name '{}' is invalid: no '-' separator found at pos {}", directory_name, pos));
+    }
+    return std::make_pair(directory_name.substr(0, pos), table_id(utils::UUID(directory_name.substr(pos + 1))));
+}
+
+void backup_task_impl::run_in_thread(sstring data_dir) {
+    std::exception_ptr ex;
+    //
+    // The keyspace data directories and their snapshots are arranged as follows:
+    //
+    //  <data dir>
+    //  |- <keyspace name1>
+    //  |  |- <column family name1>
+    //  |     |- snapshots
+    //  |        |- <snapshot name1>
+    //  |          |- <snapshot file1>
+    //  |          |- <snapshot file2>
+    //  |          |- ...
+    //  |        |- <snapshot name2>
+    //  |        |- ...
+    //  |  |- <column family name2>
+    //  |  |- ...
+    //  |- <keyspace name2>
+    //  |- ...
+    //
+    auto ks_dir = fs::path(data_dir) / _ks;
+    auto ks_dir_lister = directory_lister(ks_dir, lister::dir_entry_types::of<directory_entry_type::directory>());
+    auto close_ks_dir_lister = deferred_close(ks_dir_lister);
+
+    snap_log.trace("Scan keyspace dir {}", ks_dir.native());
+    while (auto table_ent = ks_dir_lister.get().get()) {
+        auto snapshot_dir = ks_dir/ table_ent->name / sstables::snapshots_dir / _snapshot_name;
+        auto has_snapshot = file_exists(snapshot_dir.native()).get();
+
+        snap_log.trace("Check table snapshot dir {} (exists={})", snapshot_dir.native(), has_snapshot);
+        if (!has_snapshot) {
+            continue;
+        }
+
+        auto cf_name_and_uuid = extract_cf_name_and_uuid(table_ent->name);
+        auto snapshot_dir_lister = directory_lister(snapshot_dir, lister::dir_entry_types::of<directory_entry_type::regular>());
+        auto close_snapshot_dir_lister = deferred_close(snapshot_dir_lister);
+
+        while (auto component_ent = snapshot_dir_lister.get().get()) {
+            auto gh = _uploads.hold();
+            auto component_name = snapshot_dir / component_ent->name;
+            auto destination = fmt::format("/{}/{}/{}/{}", _bucket, cf_name_and_uuid.first, _snapshot_name, component_ent->name);
+            snap_log.trace("Upload {} to {}", component_name.native(), destination);
+            std::ignore = _client->upload_file(component_name, destination).handle_exception([this, comp = component_name] (auto ex) {
+                snap_log.error("Error uploading {}: {}", comp.native(), ex);
+                _ex = std::move(ex);
+            }).finally([gh = std::move(gh)] {});
```

Comment:
This runs in the background on purpose, right?
Shouldn't we control parallelism so not to overwhelm the number of sessions?
Who waits on those (currently abandoned) futures?

Reply:
A comment explaining the the future can be ignored since it is waited on indirectly via the `_uploads` gate would be good to have.

Reply:
Parallelizm is controlled in two ways -- s3 client memory semaphore and http client maximum number of connections. I will add comments